src/resources/views/crud/fields/select_multiple.blade.php

```diff
@@ -6,6 +6,7 @@
         $options = call_user_func($field['options'], $field['model']::query());
     }
     $field['allows_null'] = $field['allows_null'] ?? true;
+    $field['value'] = isset($field['value']) && is_array($field['value']) ? $options->whereIn((new $field['model'])->getKeyName(), $field['value']) : collect();
```

Comment:
This didn't make the `select_multiple` work inside `repeatable` for me. The adding seemed to work, but then on `edit` the saved options no longer show up as selected.

Same for `select2_multiple`, where copy-pasted this fix, thinking it works.

Reply:
Could you please explain what this fixes and why it's needed @pxpm ?

Reply:
This is needed here because in that field we expect the options to be a collection of models when retrived from the pivot table in the database. Ex: `$entry->tags` return a collection of `tag model` from `monster_tag` table. In repeatable the data is saved as a single array of id's inside the json. `tags => [1,3,5]`, so we need to convert it back to the actual models getting it from the retrieved options. 

For me it breaks with `Call to a member function pluck() on array` when doing `$field['value']->pluck()` because `$field['value']` is an array of selected id's. 

But indeed the conditional is wrong, I was just testing this in `dummy` and didn't see this breaks the regular fields. I've just submitted a fixed version and fixed in `select2_multiple` too. 

https://recordit.co/eKaUao4dYb

Reply:
Awesome! I can confirm it now works for entries submitted after this PR. 

However... it still didn't work for me when I tested using OLD entries. Let me tell you the story 😀

![](https://media3.giphy.com/media/cvmQECBQy9pNC/giphy.gif)

I've:
- added an entry in Dummies on the `master` branch, with the standard Relationship > select_multiple and select2_multiple filled in;
- switched the branch to this PR, reloaded the Edit page for that entry... now the entries do not show up as selected, even though they are selected in the database:


Here's a side-by-side comparison of two entries (left added on `master`, right added on `repeatable-as-array`, both of them previewed on `repetable-as-array`):

<img width="1439" alt="Screenshot 2021-11-11 at 10 30 44" src="https://user-images.githubusercontent.com/1032474/141264645-4f49045a-253e-4442-bcf8-b0989845a1f5.png">

I've done this to test that old projects' existing entries will still "just work" after the update. From what I understand... they don't.

I did a `dump($this->crud->getCurrentEntry()->relationships)` to see if there are any differences between what is stored in the database, and what do you know, there are:

<img width="1440" alt="Screenshot 2021-11-11 at 10 30 25" src="https://user-images.githubusercontent.com/1032474/141264640-464e4e28-098c-48f3-a023-652b24dc09ab.png">

I see two big differences:
- on `master` the name is `categories[]`, but on `repeatable-as-array` it's `categories`;
- on `master` it's a JSON, but on `repeatable-as-array` it's an array;

I've done the `dump()` inside `setup()`, so it's way before it reaches the actual field blade file. 

Here's what I think:
- (1) We cannot expect people to change their DB entries, to use a new version of the `repeatable` field. It needs to work with the exact same DB entries. And when we save a new entry, what gets saved to look and work the same.
- (2) I was surprised to see that the name on `master` is `categories[]`... that's so ugly! It makes it super-difficult to use that value in your project... Seems like a huge mistake on our end...
- (3) Please investigate this and tell me what you find and let's decide what we can do about this.

Reply:
Hey @tabacitu indeed 2) is the problem. 

To make it work I added the check for `name` or `name[]` in the data we store in the database and it worked out of the box.

I pushed the fix here https://github.com/Laravel-Backpack/CRUD/pull/3905/commits/01436726c8e2a8588eadec6978de2db979f8914d

Reply:
Can confirm it's working now, thanks @pxpm 🎉🎉🎉 

---

But **I don't think we're done here**. There's still a HUGE breaking change here. Imagine this scenario:
- dev developed admin panel in Backpack 4.1, using repeatable that stores categories as `categories[]`; codes his entire front-end app based on the information that this `[]` is at the end when he gets and sets categories outside his admin panel;
- dev upgrades his admin panel to Backpack 4.2, uses `repeatable` that stores categories as `categories`; his front-end will need to be changed too, because now there's no `[]` at the end;

Don't get me wrong. I hate the `categories[]` - never knew we had it. I'd say it's a bug, even. 

But if we do something like this (make the field basically rewrite the repeatable input name from `categories[]` to `categories`), then **we also have to tell the developer about it, in the upgrade guide. We need to tell him that the field will change the data structure in his database. That wherever he's using that data, he should assume it can be either `categories[]` or `categories`.** 

This makes it much more difficult to upgrade... They don't only have to change the validation, but also every bit of code that's using something stored using the `repeatable`.

----

I never wanted this 4.2 upgrade to be difficult. On the contrary, I wanted it to be a chinch. Even for the `repeatable` validation my plan was to tell people "_you can rewrite it to something like `'your_field.*.name' => 'required'`, which is much cleaner, but if you're too lazy to do that you can just add this line in your old validation, to make it work inside the repeatable v2_":

```diff
    public function rules()
    {
        return [
            'name'        => 'required|min:5|max:255',
            'description' => 'required',
            'simple'      => function ($attribute, $value, $fail) {
+               $fieldGroups = is_array($value) ? $value : json_decode($value);

                // do not allow repeatable field to be empty
                if (count($fieldGroups) == 0) {
                    return $fail('The simple field group must have at least one item.');
                }

                // ALTERNATIVE:
                // allow repeatable field to be empty
                // if (count($fieldGroups) == 0) {
                //   return true;
                // }

                // SECOND-LEVEL REPEATABLE VALIDATION
                // run through each field group inside the repeatable field
                // and run a custom validation for it
                foreach ($fieldGroups as $key => $group) {
                    $fieldGroupValidator = Validator::make((array) $group, [
                        'text'  => 'required|min:10|max:255',
                        'email' => 'required|email',
                    ]);

                    if ($fieldGroupValidator->fails()) {
                        // return $fail('One of the entries in the '.$attribute.' group is invalid.');
                        // alternatively, you could just output the first error
                        return $fail($fieldGroupValidator->errors()->first());
                        // or you could use this to debug the errors
                            // dd($fieldGroupValidator->errors());
                    }
                }
            },
        ];
    }
```

---

So let me ask you this. **Is there any benefit from introducing this `category[]` to `category` renaming, other than the fact that it's ugly?**

(A) Do we _really_ need to do this or could we just as well keep using the ugly `category[]` to keep the breaking changes to a minimum? 

(B) Maybe we could a flag/attribute on the `repeatable` field, something like `remove_trailing_brackets`; if `true` we do what you did now and rewrite; if `false` we keep the old `category[]` behaviour;

Not sure about this... what do you think @pxpm ? Aren't you worried this is becoming a pain to upgrade? We need to walk this difficult line between not disturbing our old `repeatable` users, when upgrading... and satisfying new `repeatable` users who would like to use any fields inside it...